doc/api/esm.md

```diff
@@ -1506,9 +1507,9 @@ _isImports_, _conditions_)
 > 7. If _pjson?.type_ exists and is _"module"_, then
 >    1. If _url_ ends in _".js"_, then
 >       1. Return _"module"_.
->    2. Throw an _Unsupported File Extension_ error.
+>    2. return **undefined**.
 > 8. Otherwise,
->    1. Throw an _Unsupported File Extension_ error.
+>    1. return **undefined**.
```

Comment:
This would be a breaking change to the resolution algorithm. cc @guybedford 

I think Node could and still should throw this error, it would just be as part of resolution _and loading,_ as opposed to just resolution. So all we really need to do is rename the section “Module resolution and loading algorithm,” and you add the steps for loading and move the error to there. From the consumer’s perspective there would be no breaking changes.

Reply:
> > If we’re moving the check for unsupported schemes into load, then we should move the check for unsupported file types/extensions into load as well (or does that already happen as part of load?)

> @GeoffreyBooth already happens today

All the more reason then to extend this algorithm definition to include the loading phase, to document which errors get thrown during that phase too.

Reply:
> I agree. But I think this is out of scope for this issue.

I don’t think so, as this PR is all about moving an error into the loading phase. The loading phase is so simple that we don’t currently define that algorithm (it’s essentially “load the source from the resolved URL”) so it’s really just adding the validation checks that happen before that step.

Reply:
@GeoffreyBooth arent loaders still experimental? 
the docs on the resolve hook are pretty clear https://nodejs.org/api/esm.html#resolvespecifier-context-nextresolve
> The loaders API is being redesigned. This hook may disappear or its signature may change. Do not rely on the API described below.

so I think it should ok to land this and adjust the docs in a follow-up PR

Reply:
> arent loaders still experimental?

Loaders are experimental, but the built-in ESM Loader’s resolution algorithm is stable and has been such for years. What we document about it here is the basis for bundlers that replicate the algorithm, so it’s important that the documentation be complete. Those other tools replicate the error flows too.

Reply:
I'll write it down, as suggested by @GeoffreyBooth  (by looking at the code and trying to replicate the same "algorithm style" as in the resolver algorithm).